packages/deployments/contracts/contracts/messaging/libraries/DomainIndexer.sol

```diff
@@ -0,0 +1,160 @@
+   * @dev This hash should be re-calculated whenever the connectors array is updated.
+   */
+  bytes32 public connectorsHash;
+
+  /**
+   * @notice Shortcut to reverse lookup the index by domain. We index starting at one so the zero value can
+   * be considered invalid (see fn: `isDomainSupported`).
+   * @dev This should be updated whenever the domains array is updated.
+   */
+  mapping(uint32 => uint256) private domainToIndexPlusOne;
+
+  // ============ Getters ============
+
+  /**
+   * @notice Convenience shortcut for supported domains. Used to sanity check adding new domains.
+   * @param _domain Domain to check.
+   */
+  function isDomainSupported(uint32 _domain) public view returns (bool) {
+    return domainToIndexPlusOne[_domain] != 0;
+  }
+
+  /**
+   * @notice Gets the index of a given domain in the domains and connectors arrays.
+   * @dev Reverts if domain is not supported.
+   * @param _domain The domain for which to get the index value.
+   */
+  function getDomainIndex(uint32 _domain) public view returns (uint256) {
+    uint256 index = domainToIndexPlusOne[_domain];
+    require(index != 0, "!supported");
+    return index - 1;
+  }
+
+  /**
+   * @notice Gets the corresponding hub connector address for a given spoke domain.
+   * @dev Inefficient, should only be used by caller if they have no index reference.
+   * @param _domain The domain for which to get the hub connector address.
+   */
+  function getConnectorForDomain(uint32 _domain) public view returns (address) {
+    return connectors[getDomainIndex(_domain)];
+  }
+
+  /**
+   * @notice Validate given domains and connectors arrays are correct (i.e. they mirror what is
+   * currently saved in storage).
+   * @dev Reverts if domains or connectors do not match.
+   */
+  function validateDomains(uint32[] calldata _domains, address[] calldata _connectors) public view {
+    // Validate that given domains match the current array in storage.
+    require(keccak256(abi.encode(_domains)) == domainsHash, "!domains");
+    // Validate that given connectors match the current array in storage.
+    require(keccak256(abi.encode(_connectors)) == connectorsHash, "!connectors");
+  }
+
+  // ============ Helper Functions ============
+
+  /**
+   * @notice Handles all mutex for adding support for a given domain.
+   * @param _domain Domain for which we are adding support.
+   * @param _connector Corresponding hub connector address belonging to given domain.
+   */
+  function addDomain(uint32 _domain, address _connector) internal {
+    // Sanity check: domain does not already exist.
+    require(!isDomainSupported(_domain), "exists");
+    // Sanity check: connector is reasonable.
+    require(_connector != address(0), "!connector");
+
+    // Push domain and connector to respective arrays.
+    domains.push(_domain);
+    connectors.push(_connector);
+    // Set reverse lookup.
+    uint256 _indexPlusOne = domains.length;
+    domainToIndexPlusOne[_domain] = _indexPlusOne;
```

Comment:
why not just use the index instead of `domainToIndexPlusOne` ? or is it to avoid the `0` case altogether? (if so, we should add a comment to the mapping declaration)

Reply:
Yep, it's a sneaky shortcut to avoid 0-case so it's a dual purpose mapping. There is a comment in the mapping declaration about it, but lemme make sure it's highlighted

Reply:
oh must've missed it in the diff -- its in there, nice